website_cms/models/website_security_mixin.py

```diff
@@ -0,0 +1,232 @@
+    _admin_groups = (
+        'base.group_website_publisher',
+    )
+
+    view_group_ids = fields.Many2many(
+        string='View Groups',
+        comodel_name='res.groups',
+        help=(u"Restrict `view` access to this item to specific groups. "
+              u"No group means anybody can see it.")
+    )
+
+    def _is_admin(self):
+        for gid in self._admin_groups:
+            if self.env.user.has_group(gid):
+                return True
+        return False
+
+    def _check_user_groups(self, group_ids):
+        """Check wheter current user matches given groups."""
+        return any([x in self.env.user.groups_id._ids
+                    for x in group_ids])
+
+    def _bypass_check_permission(self):
+        """Bypass checking permissions.
+
+        Bypass if:
+
+        * you are super-user
+        * you are in a group listed in `_admin_groups`
+        """
+        if self._uid == SUPERUSER_ID:
+            return True
+        admin_groups = []
+        for k in self._admin_groups:
+            ref = self.env.ref(k)
+            if ref:
+                admin_groups.append(ref.id)
+        return self._check_user_groups(admin_groups)
+
+    @api.model
+    def check_permission(self, obj=None, mode='view'):
+        """Check permission on given object.
+
+        @param `obj`: the item to check.
+        If not `obj` is passed `self` will be used.
+
+        @param `mode`: the permission mode to check.
+        """
+        obj = obj or self
+        fname = mode + '_group_ids'
+
+        if fname not in obj._model:
+            # nothing to check here
+            return True
+
+        # obj comes w/ a temporary env,
+        # w/ uid beeing an instance of ir_http.RequestUID
+        # which is not suitable for std ORM operations.
+        # Let's make sure we get the right user here!
+        if request.session.uid:
+            _obj = obj.with_env(self.env(user=request.session.uid))
+        else:
+            _obj = obj.with_env(request.env)
+
+        if _obj._bypass_check_permission():
+            return True
+        if not _obj[fname]:
+            # no groups, fine
+            return True
+
+        check1 = _obj._check_user_groups(_obj[fname]._ids)
+        return check1
+
+    @api.model
+    def _search(self, args, offset=0, limit=None,
+                order=None, count=False, access_rights_uid=None):
+        """Implement security check based on `website_published`
+        and `view_group_ids`.
+
+        The goal is to hide items that current user cannot see
+        in the frontend (like navigation, listings)
+        and, why not, even in the backend.
+
+        Here if we find a contraint on any group
+        we filter out what the current user cannot view.
+
+        Disclaimer: yes, we could use `ir.rule`s for this
+        but since we want the end user to select a group
+        easily without having to get in touch w/ rules,
+        we chosed this way.
```

Comment:
If the group is attached to some record, then you can add a ir.rule like `[("somefield", "in", user.group_id)]` and remove lots of code